keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -589,6 +592,152 @@ associativeList.all(e, e.key1 == 'a' && e.key2 == 'b' && e.val == 100)
+  - But still allow O(n<sup>2</sup>) or worse so long as the CRD author adds a `maxLength` limit to the lists and the resulting cost is within cost bounds (e.g. O(n<sup>3</sup>) is still fine for a list size of 3).
+
+Also, not requiring `maxLength` on most O(n) CEL expressions keeps the cost system low friction; the majority 
+of CEL expressions can be written and used without bumping into cost limits or needing to set `maxLength`.
+
+For the cases where a `maxLength` is needed to ensure cost is acceptable, encouraging CRD authors to include 
+`maxLength` on variable length data elements is generally considered good hygiene anyway, so having a cost 
+system that incentivizes developers to set this field, has other benefits.
+
+We will use the minimum possible size of list elements to estimate the size (e.g. objects with many fields 
+will have much larger minimum sizes). This helps us pick smaller worst case list sizes that we might pick if we had to assume all lists contain 1 byte elements.
+
+For example, the list element size `x` in this schema:
+
+```
+spec:
+  type: object
+  properties:
+    x:
+      x-kubernetes-validation:
+      - rule: "self.all(element, element.y.startsWith('prefix:'))"
+      type: array
+      items:
+        type: object
+        properties:
+          name: y
+          type: string
+          name: z
+          type: string
+```
+
+The worst case list size of `x` would be based on the size of an object with two strings (`y` and `z`).
+
+We will also account for where in a schema CEL expression is placed when calculating its cost. For example:
+
+```
+spec:
+  type: object
+  properties:
+    x:
+      type: array
+      items:
+        type: object
+        properties:
+          name: y
+          type: string
+          x-kubernetes-validation:
+          - rule: "self.startsWith('prefix:')"
+```
+
+In this example the validation rule is contained within a list, so its expression cost would be `cost(startsWith) * worst_case_length(x)`. Note that this means some of the cost calculation will
+happen outside of the CEL cost function (which would only calculate the `startsWith` cost).
+
+During CRD validation, when an expression exceeds the cost limit, the error message will include both the limit and the cost of the expression to facilitate debugging. The cost calculation rules will be documented.
+
+#### Time limits
+
+We believe CEL cost will be sufficient to bound the resource utilization of the vast majority of CEL
+expressions. As a backstop we will use timeouts. We will set these timeouts high enough that they are
+not disruptive to all but the most extremely resource starved api-servers. We don't believe this will
+cause problems in practice because there are other timeouts that will trigger under those conditions.
+
+Per-request and per-expression execution time will be constrained via go cancellation and a context-passed timeout.
```

Comment:
I don't like that the time limit could make this inconsistent depending on how busy apiserver is. Does CEL offer a feature for counting up & limiting actual runtime cost? I'd really prefer a limit that was reproducible given the input object(s).

Reply:
agreed

Reply:
This seems doable to me. At compile time we can compute a cost heuristic with incomplete information. Summing a cost heuristic at runtime seems so harder, and maybe easier. I'd be interested in exploring this. If we can get it working, I'd be in favor of using it instead of wall clock time.

Reply:
I would strongly prefer this to a time-based approach if it is doable.

Reply:
Oh, hrmm… I wasn't thinking about calculating cost per request. The worst case pre-calculation at CRD edit time seems more palatable.

Reply:
I don't think you should compute cost-per-request, I think you should give CEL an expression budget and let it halt if the expression goes over its budget.

Reply:
A case that seems relevant to this discussion:

```
spec:
  type: object
  properties:
    x:
      type: array
      items:
        type: object
        properties:
          name: y
          type: int
          x-kubernetes-validation:
          - rule: "self < 100"
```

Here the rule is cheap (I'm pretty sure the cost is 1).  But it's contained in a list which might be huge.

Reply:
The current alternatives I'm seeing:

1. Proactive: We pre-compute the worst case costs of each expression when validation rules are written to a CRD. Expressions that are O(n) or worse will usually need to set maxLengths to bring their worst case cost down into range
2. Reactive: We count cost at CEL expression runtime, accumulating all the cost that an expression has consumed and exiting early if the cost limit is hit
3. We do both Proactive and Reactive (but with different limits?)
4. We do one of the above but also with a per-request limit (to handle the case where a CRD has many expensive CEL expressions)
5. We do one of the above but limit the total number of expressions allowed in a CRD (to handle the same situation as 4)

Reply:
I think you should keep two budgets, overall and per expression.

When running an expression, its budget is the minimum of the remaining overall budget and of the per-expression budget.

After running an expression, you should decrement whatever it used from the remaining overall budget.

(Assuming you can do this with CEL)